iXBRLViewerPlugin/viewer/src/js/textblockviewer.js

```diff
@@ -0,0 +1,36 @@
+// Copyright 2023 Workiva Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import $ from 'jquery';
+
+import { Dialog } from './dialog.js';
+
+export class TextBlockViewerDialog extends Dialog {
+    constructor() {
+        super(".dialog.text-block-viewer");
+        this.addButton("Dismiss", true);
+    }
+
+    displayTextBlock(textBlockValue) {
+        const iframe = this.node.find("iframe").get(0);
+        const doc = iframe.contentDocument || iframe.contentWindow.document;
+        doc.open();
+        const html = 
+            "<!DOCTYPE html><html><head><title></title></head><body>" 
+            + textBlockValue 
```

Comment:
Good point, except that when I try it, it seems that the viewer creation process converts empty elements into self-closing elements, so we actually have the opposite problem.    If I have text block content of:

```
Empty div <div style="color: red"></div> Is this red?
```

Then the previewer displays "is this red" in red, because the `<div></div>` has been converted to  a self closing element.  `xhtmlserializer.py` takes care of this for the document itself and will expand a self-closing `<div />`, but this happens after Arelle has extracted text block content.  I'll dig into this a bit more.

Reply:
It looks like this is in the text block extraction code [here](https://github.com/Arelle/Arelle/blob/master/arelle/XmlUtil.py#L278).  I think that will always cause empty elements to become self-closing elements. 

I'm a bit stuck on this one.  Options:

* Force the iframe to render as XHTML.  I think this would solve the problem, but I cannot find a way to do it.
* Fix Arelle to preserve self-closing vs empty tags, but I don't think that this distinction is preserved in the model.
* Change Arelle to use the same logic as the viewer when serialising XHTML (self-close tags that must be empty, always expand other tags)
* Try to fix up the XHTML at the point that we add it to the iframe (which probably means an XML parse followed by a custom serializer)
* Extract text block contents directly from the iXBRL in the viewer, rather than using the values in the JSON data.

I do think that the current behaviour in Arelle is potentially problematic, as any out-of-context rendering of a text block in HTML rather than XHTML mode will break if it encounters empty elements (e.g. empty spans or divs).

Reply:
Raised https://github.com/Arelle/Arelle/issues/574 against Arelle.

Reply:
* re: option 1, I also tried before writing my last comment, and I also couldn't find a way to do it.
* re: option 2, agreed, likely not viable.
* re: option 3, seems viable.
* re: option 4, seems viable, but probably a lot of work.  I've never parsed XML from JavaScript, but a quick test suggests `new DOMParser().parseFromString('<html><br></br></html>', 'application/xml')` works (and [caniuse](https://caniuse.com/?search=DOMParser) suggests it's usable).
* re: option 5, it doesn't feel great to duplicate XBRL value processing in ixbrl-viewer even though that's what we do today.

So, preference order would be 3 then 4 then 5.

That said, it occurs to me now that for any option 1-4 (or PR as-is), ixbrl-viewer would be relying on Arelle to transform to XHTML but on its own processing for `escaped="true"` to determine whether to (effectively) do `innerHTML`, which isn't great: if ixbrl-viewer's processing wrongly concludes a fact is encoded XML but Arelle doesn't, then we risk "benign" string facts becoming active content.  (I guess `<ix:nonNumeric escaped="false">&lt;b&gt;text&lt;/b&gt;</ix:nonNumeric>` is also technically allowed, but I guess that's unlikely, and ixbrl-viewer doesn't have to cater to it.)  I wonder if the Python should add something to the JSON (e.g., set of used concept names that are XHTML) so the JavaScript can accurately know whether to trust the value to be validated XHTML.  If the new JSON item isn't allowed, then we could fallback to the safe HEAD algorithm of extracting text nodes (for [versioning](https://github.com/Workiva/ixbrl-viewer#javascript-versioning)).

As an aside, re: option 3, another HTML writing quirk we discovered with ixbrl-viewer (without text blocks), is that for `<style>a>b{}</style>`, lxml unnecessarily escapes and writes out `<style>a&gt;b{}</style>`, which does not work in an HTML context, so I wonder if manual reserialization is the only/best answer even for ixbrl-viewer's `XHTMLSerializer`.  I wonder if the XHTML-vs-plain-HTML-rendering quirks are worth mentioning in [Designing HTML for XBRL](https://www.xbrl.org/Specification/html-for-ixbrl-wgn/WGN-2022-12-07/html-for-ixbrl-wgn-2022-12-07.html).  Well, we've also discovered that the browser uses different `document.compatMode`, so you have to be careful with CSS styles too if you want documents to render the same regardless of what MIME type users serve the generated iXBRL (e.g., whether they use `.xhtml` or `.htm`/`.html` extension, and we know that different ESEF jurisdictions have different requirements).

Reply:
> So, preference order would be 3 then 4 then 5.

Agreed.

Regarding the question of when to treat the value as XHTML, I think as per the Spec WG discussion, this is determined by the concept's datatype, but because of the disconnect between the `escape` attribute and the datatype, there's no guarantee that the resulting value is actually well-formed XML.

As things stand, I think that the viewer _can_ trust that the content is validated XHTML, because we know that it was derived by escaping the content of a valid XHTML document.  The risk is that the viewer implies that a non-text-block fact such as company name may contain XHTML and will be treated as such by consumers.  e.g. if someone tags their company name as "My **Company**" using `escape="true"`, the viewer will currently render that as "My **Company**" suggesting that it's a legitimate thing to do, rather than showing it as `My <b>Company</b>`, which is appropriate because it's a string, not a text block.

If we switch to doing this based on datatype, we should also XML (or XHTML) validate the resulting value, and display an error in the viewer if the result is invalid.

> I wonder if manual reserialization is the only/best answer even for ixbrl-viewer's XHTMLSerializer.

I was having the same thought based on the discussion on #384.

Some time ago, I went down the rabbit hole of trying to make it possible to embed JavaScript in an HTML/XHTML polyglot, which is doable using CDATA in a comment, but you need to make sure that the JS doesn't accidentally include `]]>` which is not that unlikely in minified JS.  If it does you can either escape (in a string) or modify (otherwise) the `]]>` but knowing which to do that requires knowing whether you're in a JS string literal or not, which requires a JS parser, and JS parsing is horrible thanks to the `/regex/` literal.

Fortunately, I think we can safely not support inline JS, and dealing with CSS should be more straightforward.

> I wonder if the XHTML-vs-plain-HTML-rendering quirks are worth mentioning in [Designing HTML for XBRL](https://www.xbrl.org/Specification/html-for-ixbrl-wgn/WGN-2022-12-07/html-for-ixbrl-wgn-2022-12-07.html).

I had exactly the same thought.  I'll raise a Spec WG issue.

Reply:
> If we switch to doing this based on datatype, we should also XML (or XHTML) validate the resulting value, and display an error in the viewer if the result is invalid.

That sounds reasonable.

To be sure, one concern with trying to render arbitrary iXBRL XHTML facts  rather than just `escape="true"` is that security is a bit more difficult for running ixbrl-viewer on arbitrary content, particularly for something like `<ix:nonNumeric escape="false">&lt;script>window.alert('');&lt;/script></ix:nonNumeric>`.  In that case, only ixbrl-viewer would see the `<script>`, whereas if ixbrl-viewer limits itself to facts with XHTML datatype _and_ that are `escape="true"`, then the literal `<script>` tag must be present in the original XHTML, which is likely easier for automated scanning.

Reply:
Option 3 was done in Arelle (https://github.com/Arelle/Arelle/pull/592).

I've now made Arelle flag text block facts in the JSON data, and the text block viewer is only enabled if this flag is present, and the value is escaped.  The viewer XML-validates the text blocks, and displays an error rather than the text block if invalid, although this should be impossible given that we're requiring the source is from an escaped text block.